third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -170,105 +233,219 @@ impl BorrowEdgeKind {
-    moved: SetDomain<TempIndex>,
-}
-
 impl AbstractDomain for LifetimeState {
-    /// The join operator of the dataflow analysis domain. This calls into `join_label_map`
-    /// which does the work of graph gluing.
+    /// The join operator of the dataflow analysis domain.
+    ///
+    /// Joining of lifetime states is easy for the borrow graph, as we can simply join the node representations
+    /// using the same label. This is consistent because each label is constructed from the program point.
+    /// However, if it comes to the mappings of globals/locals to labels, we need to unify distinct labels of the
+    /// two states. Consider `$t1 -> @1` in one state and `$t1 -> @2` in another state, then we need to unify
+    /// the states under labels `@1` and `@2` into one, and renamed any occurrence of the one label by the other.
     fn join(&mut self, other: &Self) -> JoinResult {
+        // Join the graph
         let mut change = self.graph.join(&other.graph);
+        self.check_graph_consistency();
+
+        // A label renaming map resulting from joining lifetime nodes.
+        let mut renaming: BTreeMap<LifetimeLabel, LifetimeLabel> = BTreeMap::new();
 
         let mut new_local_to_label_map = std::mem::take(&mut self.local_to_label_map);
         change = change.combine(self.join_label_map(
             &mut new_local_to_label_map,
             &other.local_to_label_map,
-            |temp| MemoryLocation::Local(*temp),
+            &mut renaming,
         ));
         let mut new_global_to_label_map = std::mem::take(&mut self.global_to_label_map);
         change = change.combine(self.join_label_map(
             &mut new_global_to_label_map,
             &other.global_to_label_map,
-            |id| MemoryLocation::Global(id.clone()),
+            &mut renaming,
         ));
         self.local_to_label_map = new_local_to_label_map;
         self.global_to_label_map = new_global_to_label_map;
 
+        if !renaming.is_empty() {
+            Self::rename_labels_in_graph(&renaming, &mut self.graph);
+            Self::rename_labels_in_map(&renaming, &mut self.local_to_label_map);
+            Self::rename_labels_in_map(&renaming, &mut self.global_to_label_map);
+            for old in renaming.keys() {
+                self.graph.remove(old);
+            }
+            change = JoinResult::Changed;
+        }
+        self.check_graph_consistency();
+
         change = change.combine(self.moved.join(&other.moved));
         change
     }
 }
 
 impl LifetimeState {
-    /// Joins two maps with labels in their range. For overlapping keys, a new `Skip`
-    /// node is created and configured as a child of both label nodes, gluing the
-    /// graphs together.
+    /// Joins two maps with labels in their range. For overlapping keys pointing to different labels,
+    /// the nodes behind the labels in the graph are joined, and the label in the other graph is
+    /// replaced by the given one.
```

Comment:
"... is replaced by the given one" -> "... is replaced by the given one in the `renaming` output parameter".